framework/finalizer.go

```diff
@@ -24,30 +24,57 @@ type patchSpec struct {
 
 func (f *Framework) addFinalizer(obj interface{}) (stopReconciliation bool, err error) {
 	restClient := f.k8sClient.CoreV1().RESTClient()
-	patch, path, result, err := createAddFinalizerPatch(obj, f.name)
-	if err != nil {
-		return false, microerror.Mask(err)
-	}
-	if patch == nil {
-		return result, err
-	}
-	p, err := json.Marshal(patch)
+	// We get the accessor of the object which we got passed from the framework.
+	accessor, err := meta.Accessor(obj)
 	if err != nil {
 		return false, microerror.Mask(err)
 	}
+	path := accessor.GetSelfLink()
+	// Continuing with reconciliation is set as a default here.
+	stopReconciliation = false
+
 	operation := func() error {
-		res := restClient.Patch(types.JSONPatchType).AbsPath(path).Body(p).Do()
-		if res.Error() != nil {
+		// We get an up to date version of our object from k8s.
+		res := restClient.Get().AbsPath(path).Do()
+		// We parse the response from the RESTClient to runtime object.
+		obj, err = res.Get()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		patch, result, err := createAddFinalizerPatch(obj, f.name)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if patch == nil {
+			// patch is empty, nothing to do.
+			// We trust createAddFinalizerPatch with the decision to stop reconciliation.
+			stopReconciliation = result
+			return nil
+		}
+		p, err := json.Marshal(patch)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		res = restClient.Patch(types.JSONPatchType).AbsPath(path).Body(p).Do()
```

Comment:
Looks like `res` is not used for anything but getting the error. We can just call `Error()` here at the end and only deal with that. Would be more natural to read and use below.

Reply:
This is fixed.